zebra-chain/src/transaction/tests/vectors.rs

```diff
@@ -549,7 +555,8 @@ fn test_vec243_1() -> Result<()> {
 
     let transaction = ZIP243_1.zcash_deserialize_into::<Transaction>()?;
 
-    let hasher = SigHasher::new(&transaction, NetworkUpgrade::Sapling, &[]);
+    let hasher = SigHasher::new(&transaction, NetworkUpgrade::Sapling, Arc::new(Vec::new()))
```

Comment:
Optional/Nitpick: This could be simplified if we accept `impl Into<Vec<Output>>` for the `all_previous_outputs` param in `SigHasher::new()`.

```suggestion
    let hasher = SigHasher::new(&transaction, NetworkUpgrade::Sapling, [])
```

```rust

    pub fn new(
        trans: &Transaction,
        nu: NetworkUpgrade,
        all_previous_outputs: impl Into<Vec<transparent::Output>>,
    ) -> Result<Self, Error> {
        Ok(SigHasher {
            precomputed_tx_data: PrecomputedTxData::new(
                trans,
                nu,
                Arc::new(all_previous_outputs.into()),
            )?,
        })
    }

```

Reply:
Oof that would have saved me some work :rofl: I'll do it anyway, it looks much better, thanks!

Reply:
Now that I thought about it I'm not sure, the whole point of using `Arc` was to avoid cloning the entire vector. However if we do this, `CachedFfiTransaction` will have to clone it when calling `Sighasher::new()`.

I imagine there might be some trait magic to make this (make a function receive either T or `Arc<T>` and automatically wrap in a `Arc` if it is `T`) work but I couldn't find it.

In hindsight I think just moving the vector all the way down to `PrecomputedTx` might have been the best approach, but I guess it doesn't really matter that much

Reply:
> Now that I thought about it I'm not sure, the whole point of using Arc was to avoid cloning the entire vector.

It'll still do that, `Sighasher::new()` is _now_ only called once in the transaction verifier via `CachedFfiTransaction::new()`

Reply:
> Not very important here either way, I mostly wanted to advertise the pattern.

It would be nice to get rid of the duplicate field, but accepting the generic param is still unimportant.

Reply:
> We could remove the `all_previous_outputs` field altogether and make it a function that gets it from the `Sighasher` to avoid that:

That would work for `CachedFfiTransaction`, but `Sighasher` would need to do the same since `PrecomputedTxData` (a `Sighasher` field) also needs `all_previous_output`... and there is now a bit trickier, because it's put into a `MapTransparent` struct which is passed to `.map_authorization()` but I'm not sure if you can ever fetch it back after that. It just seemed simpler to use `Arc`